vowpalwabbit/cbify.cc

```diff
@@ -126,15 +115,38 @@ void finish_cbify_reg(cbify_reg& data, std::ostream* trace_stream)
   data.cb_cont_label.costs.delete_v();  // todo: instead of above
 }
 
-void copy_example_to_adf(cbify& data, example& ec)
+void cbify_adf_data::init_adf_data(const size_t num_actions, std::vector<std::vector<namespace_index>>& interactions)
 {
-  cbify_adf_data& adf_data = data.adf_data;
-  const uint64_t ss = data.all->weights.stride_shift();
-  const uint64_t mask = data.all->weights.mask();
+  this->num_actions = num_actions;
+
+  ecs.resize(num_actions);
+  for (size_t a = 0; a < num_actions; ++a)
+  {
+    ecs[a] = VW::alloc_examples(CB::cb_label.label_size, 1);
+    auto& lab = ecs[a]->l.cb;
+    CB::cb_label.default_label(&lab);
+    ecs[a]->interactions = &interactions;
+  }
+}
 
-  for (size_t a = 0; a < adf_data.num_actions; ++a)
+cbify_adf_data::~cbify_adf_data()
+{
+  for (size_t a = 0; a < ecs.size(); ++a)
   {
-    auto& eca = *adf_data.ecs[a];
+    ecs[a]->pred.a_s.delete_v();
+    VW::dealloc_example(CB::cb_label.delete_label, *ecs[a]);
+    free_it(ecs[a]);
+  }
+}
+
+void cbify_adf_data::copy_example_to_adf(parameters& weights, example& ec)
+{
```

Comment:
Rather than depending on the entire `parameters` object, can this instead just depend on the mask and stride shift?

Reply:
this might be good to do on its own PR, since I see the pattern of extracting mask and stride shift in multiple places